### PR TITLE
V0 Test Suite for Patch Tracker; + Beginnings of Test Suite for SAMA4D Sparse & Integrated

### PR DESCRIPTION
**Pull request by georgematheos**
The core addition of this PR is the subdirectory `tests/sama4d/`, containing the beginnings of an evaluation suite for SAMA4D.  See `tests/README.md` for more information.

The `tests/sama4d/video_to_tracks/` subdirectory is the most built out part of this suite at the moment; this subdirectory contains an evaluation for the patch tracker, consisting of a Task description, computation of metrics for measuring patch tracker accuracy (in the 2D keypoint tracks), assertion testing for quality based on these metrics, and visualization functionality.  This subdirectory contains some notebooks to visualize all the available test scenes.

The `tests/sama4d` directory also contains WIP subdirectories `video_to_tracks_and_segmentation` containing a WIP Task declaration and `tracks_to_segmentation` also containing a WIP Task declaration.  If the group likes these task declarations, we can turn these into first evaluations for the integrated SAMA4D dense+sparse inference algorithm, and the sparse keypoint segmentation module.

To support this evaluation code I added `tests/common/task.py`, a lightweight base class for task declarations, based on the work that @karen-sy and I did with the group a couple months ago.

This PR also adds some functionality to the `FeatureTrackData` type and modifies the patch tracking/dense model code somewhat.

Additionally, I added a Python file `tests/sama4d/data_curation.py` which constructs a list of scenes to use for the SAMA4D evaluation.  This currently includes several unity scenes, and the rotating cheez-it box scene (constructed via a Python function).  Others (@esli999, @tiffanyluong possibly of interest!) are invited to add more test scenes to this list.

### CODE REVIEW

File: b3d/io/feature_track_data.py [hunk 1]
```diff
@@ -228,6 +227,80 @@ def slice_time(self, start_frame=None, end_frame=None):
...
+            object_assignments=self.object_assignments,
+            camera_position=self.camera_position,
+            camera_quaternion=self.camera_quaternion,
+            camera_intrinsics=Intrinsics.from_array(self.intrinsics).downscale(factor).as_array()
+        )
+
+    def has_depth_channel(self):
+        return jnp.any(self.rgbd_images[..., 3] > 0.)
+    
+    def remove_points_invisible_at_frame0(self):
```
Comment (mirkoklukas) [inline b3d/io/feature_track_data.py:254]: what is this used for? does it have to live on the data class?

File: b3d/io/feature_track_data.py [hunk 2]
```diff
@@ -228,6 +227,80 @@ def slice_time(self, start_frame=None, end_frame=None):
...
+            object_assignments=self.object_assignments[mask] if self.object_assignments is not None else None,
+            camera_position=self.camera_position,
+            camera_quaternion=self.camera_quaternion,
+            camera_intrinsics=self.camera_intrinsics
+        )
+
+    def all_points_visible_at_frame0(self):
+        return jnp.all(self.keypoint_visibility[0])
+
+    def sparsify_points_to_minimum_2D_distance_at_frame0(self, min_2D_distance):
```
Comment (mirkoklukas) [inline b3d/io/feature_track_data.py:275]: same as above...

File: b3d/io/feature_track_data.py [hunk 3]
```diff
@@ -228,6 +227,80 @@ def slice_time(self, start_frame=None, end_frame=None):
...
+            rgbd_images=self.rgbd_images,
+            latent_keypoint_positions=self.latent_keypoint_positions[:, valid_indices] if self.latent_keypoint_positions is not None else None,
+            latent_keypoint_quaternions=self.latent_keypoint_quaternions[:, valid_indices] if self.latent_keypoint_quaternions is not None else None,
+            object_assignments=self.object_assignments[valid_indices] if self.object_assignments is not None else None,
+            camera_position=self.camera_position,
+            camera_quaternion=self.camera_quaternion,
+            camera_intrinsics=self.camera_intrinsics
+        )
+    
+    def min_2D_distance_at_frame0(self):
```
Comment (mirkoklukas) [inline b3d/io/feature_track_data.py:296]: same as above ...

File: b3d/io/feature_track_data.py (hunk 1 above)
Comment (mirkoklukas) [reply b3d/io/feature_track_data.py:254]: no, all good, i don't have a strong opinion, was just wondering where those methods are used.
